recipe/patches/0009-add-GRPC_DLL-mechanism-for-missing-symbols-on-window.patch

```diff
@@ -0,0 +1,163 @@
+   target_link_libraries(grpc_unsecure PUBLIC "-framework CoreFoundation")
+ endif()
+ 
++if(WIN32 AND MSVC AND BUILD_SHARED_LIBS)
++  # More details in include/grpc/impl/codegen/port_platform.h
++  target_compile_definitions(grpc_unsecure INTERFACE gRPC_DLL_IMPORTS)
++endif()
++
+ foreach(_hdr
+   include/grpc/byte_buffer.h
+   include/grpc/byte_buffer_reader.h
+@@ -4218,6 +4232,8 @@ if(WIN32 AND MSVC)
+   set_target_properties(upb PROPERTIES COMPILE_PDB_NAME "upb"
+     COMPILE_PDB_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}"
+   )
++  # More details in include/grpc/impl/codegen/port_platform.h
++  set_target_properties(upb PROPERTIES DEFINE_SYMBOL "gRPC_DLL_EXPORTS")
+   if(gRPC_INSTALL)
+     install(FILES ${CMAKE_CURRENT_BINARY_DIR}/upb.pdb
+       DESTINATION ${gRPC_INSTALL_LIBDIR} OPTIONAL
+diff --git a/include/grpc/impl/codegen/port_platform.h b/include/grpc/impl/codegen/port_platform.h
+index ae0e35cfb7..63baf21c6a 100644
+--- a/include/grpc/impl/codegen/port_platform.h
++++ b/include/grpc/impl/codegen/port_platform.h
+@@ -89,6 +89,25 @@
+ #undef GRPC_NOMINMAX_WAS_NOT_DEFINED
+ #undef NOMINMAX
+ #endif /* GRPC_WIN32_LEAN_AND_MEAN_WAS_NOT_DEFINED */
++
++// GPRC_DLL
++// inspired by https://github.com/abseil/abseil-cpp/blob/20220623.1/absl/base/config.h#L730-L747
++//
++// When building gRPC as a DLL, this macro expands to `__declspec(dllexport)`
++// so we can annotate symbols appropriately as being exported. When used in
++// headers consuming a DLL, this macro expands to `__declspec(dllimport)` so
++// that consumers know the symbol is defined inside the DLL. In all other cases,
++// the macro expands to nothing.
++// Note: gRPC_DLL_EXPORTS is set in CMakeLists.txt when building shared grpc{,_unsecure}
++//       gRPC_DLL_IMPORTS is set by us as part of the interface for consumers of the DLL
++#if defined(gRPC_DLL_EXPORTS)
++#define GRPC_DLL __declspec(dllexport)
++#elif defined(gRPC_DLL_IMPORTS)
++#define GRPC_DLL __declspec(dllimport)
++#else
++#define GRPC_DLL
++#endif // defined(gRPC_DLL_EXPORTS)
++
+ #endif /* defined(_WIN64) || defined(WIN64) || defined(_WIN32) || \
+           defined(WIN32) */
+ 
+diff --git a/src/core/lib/gprpp/fork.h b/src/core/lib/gprpp/fork.h
+index d6ef16130a..46b067c140 100644
+--- a/src/core/lib/gprpp/fork.h
++++ b/src/core/lib/gprpp/fork.h
+@@ -35,7 +35,7 @@ class ExecCtxState;
+ class ThreadState;
+ }  // namespace internal
+ 
+-class Fork {
++class GRPC_DLL Fork {
+  public:
+   typedef void (*child_postfork_func)(void);
+ 
+diff --git a/third_party/upb/upb/port_def.inc b/third_party/upb/upb/port_def.inc
+index 2b240ff6e8..dd71d745e9 100644
+--- a/third_party/upb/upb/port_def.inc
++++ b/third_party/upb/upb/port_def.inc
+@@ -259,3 +259,12 @@ void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
+ #else
+ #define UPB_TREAT_PROTO2_ENUMS_LIKE_PROTO3 0
+ #endif
++
++// More details in include/grpc/impl/codegen/port_platform.h
++#if defined(gRPC_DLL_EXPORTS)
```

Comment:
You need to change this to `UPB_DLL_EXPORTS` and similar

Reply:
Are we not circumventing the definition of CMake-magic symbols by using `DEFINE_SYMBOL`? That symbol should be there (we're explicitly setting it so), so I'm not sure what this would change?

In symmetry with the other things in the patch, it seems more likely to me that we're missing:
``` 
if(WIN32 AND MSVC AND BUILD_SHARED_LIBS)
  # More details in include/grpc/impl/codegen/port_platform.h
  target_compile_definitions(upb INTERFACE gRPC_DLL_IMPORTS)
endif()
```

Reply:
you are using the same symbol in upb.dll and grpc.dll. When building upb.dll you need `dllexport` and when building grpc.dll you need `dllimport`. With the scheme here, you can't build properly

Reply:
OK, that makes sense, thanks for the pointer 🙏